src/adapter/src/client.rs

```diff
@@ -235,9 +247,59 @@ pub struct SessionClient {
     cancel_tx: Arc<watch::Sender<Canceled>>,
     cancel_rx: watch::Receiver<Canceled>,
     timeouts: Timeout,
+    segment_client: Option<mz_segment::Client>,
+    environment_id: EnvironmentId,
 }
 
 impl SessionClient {
+    /// Parses a SQL expression, reporting failures as a telemetry event if
+    /// possible.
+    pub fn parse(
+        &self,
+        sql: &str,
+    ) -> Result<Result<Vec<Statement<Raw>>, ParserStatementError>, String> {
+        match mz_sql::parse::parse_with_limit(sql) {
+            Ok(Err(e)) => {
+                self.track_statement_parse_failure(&e);
+                Ok(Err(e))
+            }
+            r => r,
+        }
+    }
+
+    fn track_statement_parse_failure(&self, parse_error: &ParserStatementError) {
+        let session = self.session.as_ref().expect("session invariant violated");
+        let Some(user_id) = session.user().external_metadata.as_ref().map(|m| m.user_id) else {
+            return;
+        };
+        let Some(segment_client) = &self.segment_client else {
+            return;
+        };
+        let Some(statement_kind) = parse_error.statement else {
+            return;
+        };
+        let Some((action, object_type)) = telemetry::analyze_statement(statement_kind) else {
+            return;
+        };
+        let event_type = StatementFailureType::ParseFailure;
+        let event_name = format!(
+            "{} {} {}",
+            object_type.as_title_case(),
+            action.as_title_case(),
+            event_type.as_title_case(),
+        );
+        segment_client.environment_track(
+            &self.environment_id,
+            session.application_name(),
+            user_id,
+            event_name,
+            json!({
+                "statement_kind": statement_kind,
+                "error": &parse_error.error,
```

Comment:
@arusahni this is the one thing to look closely at. I snipped out the object type and action here because they're in the event name, and for other Segment events we don't duplicate information between the name and the properties. Also I removed the `details` top-level key since it seemed cleaner to have `statement_kind` and `error` be top-level keys. Does that work for you, or was there something downstream expecting a particular structure?

Reply:
I'm fine with dropping the object type and action. I had nested these under `details` to maintain symmetry with the successful DDL segment events (e.g.,

```json
{
  "context": {
    "group_id": "83cd79eb-9f37-4b98-bfc6-aba8cd05c307",
    "library": {
      "name": "unknown",
      "version": "unknown"
    }
  },
  "event": "Materialized View Created",
  "integrations": {},
  "messageId": "api-2Rbf7WnwgBXdqKYqhg36Q5CAwLx",
  "originalTimestamp": "2023-06-23T12:32:14.094003901Z",
  "properties": {
    "application_name": "psql",
    "cloud_provider": "local",
    "cloud_provider_region": "az1",
    "details": {
      "database": "materialize",
      "id": "u3",
      "item": "foo4",
      "schema": "public"
    },
    "event_source": "environment"
  },
  "receivedAt": "2023-06-23T12:32:14.132Z",
  "sentAt": null,
  "timestamp": "2023-06-23T12:32:14.094Z",
  "type": "track",
  "userId": "161bba05-7d91-4f90-b2ae-72c10b5fa988",
  "writeKey": "qSIu0LfPn2sOg8AmCWcfhgLXCjpr1bsi"
}
```
).

Reply:
Ah, I see! The reason we use `details` there is because it reflects the `mz_audit_events.details` column exactly: https://materialize.com/docs/sql/system-catalog/mz_catalog/#mz_audit_events

The `event_type`, `object_type`, and `user` fields are all reflected elsewhere in the Segment event so we don't include them in the event's properties. But if we were to include `id` or a future `application_name` column from the `mz_audit_events` table, we'd make those sibling properties of the `details` field. 

(Aside: now that I think about it, we probably ought to be including `occurred_at` in the audit Segment event rather than relying on the Segment event's timestamp, since the `occurred_at` timestamp is the timestamp oracle's timestamp and has additional properties that could be interesting to analytics, vs just the wall clock time collected by Segment.)